File: CalculatorUsingStack.c
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>

#define MAX 100

struct Stack {
    int top;
    int items[MAX];
};

// Function to create a stack
struct Stack* createStack() {
    struct Stack* stack = (struct Stack*)malloc(sizeof(struct Stack));
    stack->top = -1;
    return stack;
}

// Function to push an item onto the stack
void push(struct Stack* stack, int item) {
    if (stack->top < MAX - 1) {
        stack->items[++stack->top] = item;
    }
}

// Function to pop an item from the stack
int pop(struct Stack* stack) {
    if (stack->top >= 0) {
        return stack->items[stack->top--];
    }
    return -1; // Error code
}

// Function to peek the top item of the stack
int peek(struct Stack* stack) {
    if (stack->top >= 0) {
        return stack->items[stack->top];
    }
    return -1; // Error code
}

// Function to check if the stack is empty
int isEmpty(struct Stack* stack) {
    return stack->top == -1;
}

// Function to get the precedence of operators
int precedence(char op) 
{
    switch (op) {
        case '+':
        case '-':
            return 1;
        case '*':
        case '/':
            return 2;
    }
    return 0;
}

// Function to perform arithmetic operations
int applyOp(int a, int b, char op) {
    switch (op) {
        case '+': return a + b;
        case '-': return a - b;
        case '*': return a * b;
        case '/': return a / b;
    }
    return 0;
}

// Function to convert infix expression to postfix expression
void infixToPostfix(char* exp, char* postfix) {
    struct Stack* ops = createStack();
    int k = 0;
    for (int i = 0; exp[i]; i++) {
        if (isdigit(exp[i])) {
            postfix[k++] = exp[i];
        } else if (exp[i] == '(') {
            push(ops, exp[i]);
        } else if (exp[i] == ')') {
            while (!isEmpty(ops) && peek(ops) != '(') {
                postfix[k++] = pop(ops);
            }
            pop(ops); // Pop '('
        } else {
            while (!isEmpty(ops) && precedence(peek(ops)) >= precedence(exp[i])) {
                postfix[k++] = pop(ops);
            }
            push(ops, exp[i]);
        }
    }
    while (!isEmpty(ops)) {
        postfix[k++] = pop(ops);
    }
    postfix[k] = '\0';
    free(ops);
}

// Function to evaluate the postfix expression
int evaluatePostfix(char* postfix) {
    struct Stack* values = createStack();
    for (int i = 0; postfix[i]; i++) {
        if (isdigit(postfix[i])) {
            push(values, postfix[i] - '0');
        } else {
            int val2 = pop(values);
            int val1 = pop(values);
            push(values, applyOp(val1, val2, postfix[i]));
        }
    }
    return pop(values);
}

int main() {
    char infix[MAX], postfix[MAX];
    
    printf("Enter infix expression: ");
    fgets(infix, sizeof(infix), stdin);
    infix[strcspn(infix, "\n")] = 0; // Remove newline character

    infixToPostfix(infix, postfix);
    printf("Postfix expression: %s\n", postfix);

    int result = evaluatePostfix(postfix);
    printf("Result: %d\n", result);

    return 0;
}

File: calculatorusingstack.py
class Stack:
    def __init__(self):
        self.items = []

    def push(self, item):
        self.items.append(item)

    def pop(self):
        return self.items.pop() if not self.is_empty() else None

    def peek(self):
        return self.items[-1] if not self.is_empty() else None

    def is_empty(self):
        return len(self.items) == 0


def precedence(op):
    if op in ('+', '-'):
        return 1
    if op in ('*', '/'):
        return 2
    return 0


def apply_op(a, b, op):
    if op == '+':
        return a + b
    elif op == '-':
        return a - b
    elif op == '*':
        return a * b
    elif op == '/':
        return a // b  # Integer division like C
    return 0


def infix_to_postfix(expression):
    stack = Stack()
    postfix = []
    i = 0

    while i < len(expression):
        char = expression[i]

        if char.isdigit():
            # Handle multi-digit numbers
            num = []
            while i < len(expression) and expression[i].isdigit():
                num.append(expression[i])
                i += 1
            postfix.append(''.join(num))
            continue
        elif char == '(':
            stack.push(char)
        elif char == ')':
            while not stack.is_empty() and stack.peek() != '(':
                postfix.append(stack.pop())
            stack.pop()  # Pop '('
        elif char in "+-*/":
            while not stack.is_empty() and precedence(stack.peek()) >= precedence(char):
                postfix.append(stack.pop())
            stack.push(char)
        i += 1

    while not stack.is_empty():
        postfix.append(stack.pop())

    return postfix


def evaluate_postfix(postfix):
    stack = Stack()

    for token in postfix:
        if token.isdigit():
            stack.push(int(token))
        else:
            b = stack.pop()
            a = stack.pop()
            result = apply_op(a, b, token)
            stack.push(result)

    return stack.pop()


def main():
    infix = input("Enter infix expression: ").strip().replace(" ", "")
    postfix_list = infix_to_postfix(infix)
    print("Postfix expression:", ' '.join(postfix_list))
    result = evaluate_postfix(postfix_list)
    print("Result:", result)


if __name__ == "__main__":
    main()
